### PR TITLE
[9.0][IMP] account_banking_mandate: Improve / fixes

### PR DESCRIPTION
**Pull request by carlosdauden**
Adds a compute field in res.partner to search first valid mandate respecting partner banks order.
Fix payment_mode_id_change to set valid mandate if required
 
@Tecnativa

### CODE REVIEW

File: account_banking_mandate/models/res_partner.py [hunk 1]
```diff
@@ -11,6 +12,10 @@ class ResPartner(models.Model):
     mandate_count = fields.Integer(
         compute='_compute_mandate_count', string="Number of Mandates",
         readonly=True)
+    valid_mandate_id = fields.Many2one(
+        comodel_name='account.banking.mandate',
+        compute='compute_valid_mandate_id',
+        string='Valid Mandate')
```
Comment (pedrobaeza) [inline account_banking_mandate/models/res_partner.py]: First Valid Mandate

File: account_banking_mandate/models/res_partner.py [hunk 2]
```diff
@@ -21,3 +26,17 @@ def _compute_mandate_count(self):
             for mandate in mandate_data])
         for partner in self:
             partner.mandate_count = mapped_data.get(partner.id, 0)
+
+    @api.multi
+    def compute_valid_mandate_id(self):
+        # Dict to reduce impact with "bug" that process all partners related
```
Comment (pedrobaeza) [inline account_banking_mandate/models/res_partner.py]: `Dict for reducing the duplicated searches on parent/child partners`

File: account_banking_mandate/models/res_partner.py [hunk 3]
```diff
@@ -21,3 +26,17 @@ def _compute_mandate_count(self):
...
+        mandates_dic = {}
+        for partner in self:
+            commercial_partner_id = partner.commercial_partner_id
```
Comment (pedrobaeza) [inline account_banking_mandate/models/res_partner.py]: Name it `commercial_partner`

File: account_banking_mandate/models/res_partner.py [hunk 4]
```diff
@@ -21,3 +26,17 @@ def _compute_mandate_count(self):
...
+            if commercial_partner_id in mandates_dic:
+                partner.valid_mandate_id = mandates_dic[commercial_partner_id]
+            else:
+                mandate_id = partner.commercial_partner_id.bank_ids.mapped(
```
Comment (pedrobaeza) [inline account_banking_mandate/models/res_partner.py]: This  is incorrect if you have more than one bank account or more than one valid mandate. It should be:
```
mandate_id = partner.commercial_partner_id.mapped('bank_ids.mandate_ids')..filtered(lambda x: x.state == 'valid')[:1].id
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
We need to dig deeper in the onchange tough

**Comment by rafaelbn** [comment]
Hi @pedrobaeza , @carlosdauden attend you comments. Please review again. Thanks

**Comment by pedrobaeza** [comment]
@carlosdauden is this still relevant?

**Pull request closed without merging**